[WebGPU] webgpu:api,operation,buffers,* is not passing

mwyrzykowski opened the pull request:
#### 98da47e2740e302d72a02213c42ad421a9c871a7
<pre>
[WebGPU] webgpu:api,operation,buffers,* is not passing
<a href="https://bugs.webkit.org/show_bug.cgi?id=264825">https://bugs.webkit.org/show_bug.cgi?id=264825</a>
<a href="https://rdar.apple.com/118406537">rdar://118406537</a>

Reviewed by Dan Glastonbury and Yusuke Suzuki.

The WebGPU specification prevents ArrayBuffers returned from getMappedRange from
being transferred via postMessage calls.

* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map_ArrayBuffer-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map_detach-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map_oom-expected.txt:
Add passing expectations.

* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::ArrayBuffer::transferTo):
Use helper.

* Source/JavaScriptCore/runtime/ArrayBuffer.h:
(JSC::ArrayBuffer::isDetachable const):
Add helper member function to determine if an ArrayBuffer can be detached.

* Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
(WebCore::makeArrayBuffer):
(WebCore::GPUBuffer::internalUnmap):
Pin and unpin the buffer to prevent transferring.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::SerializedScriptValue::create):
Throw a type error if we try to transfer an ArrayBuffer which can not
be detached.

Canonical link: <a href="https://commits.webkit.org/270882@main">https://commits.webkit.org/270882@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ce5e13a1c66a58f003e0679d9c65b59572705c6d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26698 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27938 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28910 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24410 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27143 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7133 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2708 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24323 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26960 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4124 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22919 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3631 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3664 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23911 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29395 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23252 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24338 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24311 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29939 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25890 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3705 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1899 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27838 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5154 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33339 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6411 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4175 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7210 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4058 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg reviewed (approved):
:shipit: as long as JSC is happy with the `isDetactable()` inline helper.

mwyrzykowski commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -5260,7 +5260,7 @@ ExceptionOr<Ref<SerializedScriptValue>> SerializedScriptValue::create(JSGlobalOb
             if (arrayBuffer->isDetached() || arrayBuffer->isShared())
                 return Exception { ExceptionCode::DataCloneError };
-            if (arrayBuffer->isLocked()) {
+            if (!arrayBuffer->isDetachable()) {
```
Alternative to this change is calling `pinAndLock` and then adding an `unlock` to ArrayBuffer. Then change to SerializedScriptValue.cpp can be reverted.

Any opinions on which approach to take @WebKit/jsc-reviewers ?

Constellation replied:
I think this `isDetachable` looks fine.

Constellation commented on Source/JavaScriptCore/runtime/ArrayBuffer.h:
```diff
@@ -400,6 +401,11 @@ void ArrayBuffer::unpin()
 
+bool ArrayBuffer::isDetachable() const
+{
+    return !m_pinCount && !m_locked;
```
Let's also check `isShared()` too. Shared one is also not-detachable.

mwyrzykowski replied:
I'll correct this before merging, thank you

Constellation reviewed (approved):
r=me with comments.

mwyrzykowski committed:
[WebGPU] webgpu:api,operation,buffers,* is not passing
https://bugs.webkit.org/show_bug.cgi?id=264825
rdar://118406537

Reviewed by Dan Glastonbury and Yusuke Suzuki.

The WebGPU specification prevents ArrayBuffers returned from getMappedRange from
being transferred via postMessage calls.

* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map_ArrayBuffer-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map_detach-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map_oom-expected.txt:
Add passing expectations.

* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::ArrayBuffer::transferTo):
Use helper.

* Source/JavaScriptCore/runtime/ArrayBuffer.h:
(JSC::ArrayBuffer::isDetachable const):
Add helper member function to determine if an ArrayBuffer can be detached.

* Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
(WebCore::makeArrayBuffer):
(WebCore::GPUBuffer::internalUnmap):
Pin and unpin the buffer to prevent transferring.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::SerializedScriptValue::create):
Throw a type error if we try to transfer an ArrayBuffer which can not
be detached.

Canonical link: https://commits.webkit.org/270882@main
